webui/react/src/hooks/useInputEscape.ts

```diff
@@ -0,0 +1,270 @@
+  };
+
+  const onFocus = () => {
+    setFocused(true);
+    setBlurred(false);
+  };
+  return { inputRef, onBlur: handleBlur, onFocus };
+};
+
+export const useInputEscape = <T>(
+  ref?: React.ForwardedRef<T>,
+  onBlur?: <T extends HTMLInputElement | HTMLTextAreaElement>(
+    e: React.FocusEvent<T> | React.KeyboardEvent<T>,
+    previousValue?: string,
+    tagID?: string,
+  ) => void,
+): InputTextEscape => {
+  const inputRef = React.createRef<AntdInputRef>();
+  const [focused, setFocused] = useState(false);
+  const [blurred, setBlurred] = useState(false);
+
+  useImperativeHandle(ref, () => inputRef?.current as T);
+
+  useEffect(() => {
+    const handleEsc = (event: KeyboardEvent) => {
+      onEsc(focused, inputRef, event, setFocused);
+    };
+    const handleClick = (event: MouseEvent) => {
+      onClick(blurred, focused, inputRef, event, setFocused);
+    };
+    document.addEventListener('keydown', handleEsc, true);
+    document.addEventListener('click', handleClick, true);
+    return () => {
+      document.removeEventListener('keydown', handleEsc, true);
+      document.removeEventListener('click', handleClick, true);
+    };
+  }, [focused, inputRef, blurred]);
+
+  const handleBlur = <T extends HTMLInputElement | HTMLTextAreaElement>(
+    e: React.FocusEvent<T> | React.KeyboardEvent<T>,
+    previousValue?: string,
+  ) => {
+    setBlurred(true);
+    if (onBlur) onBlur(e, previousValue);
+  };
+
+  const onFocus = () => {
+    setFocused(true);
+    setBlurred(false);
+  };
+  return { inputRef, onBlur: handleBlur, onFocus };
+};
+
+export const useSelectEscape = (
+  containerRef: RefObject<HTMLDivElement>,
+  isOpen: boolean,
+  ref?: React.ForwardedRef<RefSelectProps>,
+  onBlur?: (
+    e: React.FocusEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>,
+    previousValue?: string,
+    tagID?: string,
+  ) => void,
+): SelectEscape => {
+  const inputRef = React.createRef<RefSelectProps>();
+  const [focused, setFocused] = useState(false);
+  const [blurred, setBlurred] = useState(false);
+  const [hasOpened, setHasOpened] = useState(false);
+
+  useImperativeHandle(ref, () => inputRef.current as RefSelectProps);
+
+  useEffect(() => {
+    // By the time a click event is captured in an
+    // event handler the Select will have otherwise already
+    // become unfocused and isOpen will be false.
+    // hasOpened is used in place of "isOpen" to check
+    // if the select is still open at the time of the click
+    // event.
+    if (isOpen) {
+      setHasOpened(true);
+    }
+  }, [isOpen]);
+
+  useEffect(() => {
+    const handleClick = (event: MouseEvent) => {
+      onClickSelect(event, isOpen, hasOpened, setHasOpened);
+    };
+
+    const handleEsc = (event: KeyboardEvent) => {
+      onEsc(focused, inputRef, event, setFocused);
+    };
+
+    containerRef.current?.addEventListener('keydown', handleEsc);
+    window.addEventListener('click', handleClick, true);
```

Comment:
q: why are these event listeners on different nodes?

Reply:
I added a comment here. The short answer is that how focus, blur,  the isOpen state, and how the focus is shifted back to the parent after a select closes is nuanced. And these handlers were specifically implemented to handle that nuance. So the same handler will not work if attached to the window. A new handler would need to be written to work properly when attached to the `window`.